### PR TITLE
More Kronecker Delta simplification based on evaluation identity

### LINKED ISSUE #18879: More Kronecker Delta simplification based on evaluation identity
Some kronecker delta simplification was already implemented in #17215
But I think that sympy lacks a lot of identities like 
`KroneckerDelta(n, 0) * cos(n) == KroneckerDelta(n, 0)`,
`KroneckerDelta(n, 2) * binomial(n, 2) == KroneckerDelta(n, 2)`.

I would suggest that the evaluation of the function at the specified point of Kronecker delta can generalize the simplification in #17215. But let's see if this works without much difficulty.

### PR DESCRIPTION
**Pull request by Arpan612**
More Kronecker Delta simplification based on evaluation identity
#### References to other Issues or PRs

Fixes #18879 

#### Brief description of what is fixed or changed
Trigonometric and binomial tests added for Kronecker Delta based on evaluation identity.

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by Arpan612** [comment]
@sylee957 I am not aware of any tests other than the ones already implemented. If any more tests are required, do tell me.

**Comment by Smit-create** [comment]
Adding just tests does not require release note-entry. I should be NO ENTRY

**Comment by oscarbenjamin** [comment]
This does not fix the linked issue as there is no code to do the simplification

**Comment by Arpan612** [comment]
@oscarbenjamin Sorry sir, didn't get you...Could you please elaborate?

**Comment by oscarbenjamin** [comment]
This issue is marked as "fixes #18879" meaning that if it is merged then that issue will be closed as fixed. However this does not fix that issue since it does not add the simplification suggested there so that issue should not be closed if this is merged.

**Comment by Arpan612** [comment]
Okay sir. Working on it.

**Comment by Arpan612** [comment]
@oscarbenjamin @sylee957 I am able to implement the basic syntax of piecewise. However, I am unable to understand how to represent the multiplication of two quantities here. Could you help?

**Comment by oscarbenjamin** [comment]
I am not sure that I understand your question.

In sympy a product is represented as an instance of `Mul`:
```
In [1]: 2*x                                                                                                                       
Out[1]: 2⋅x

In [2]: type(2*x)                                                                                                                 
Out[2]: sympy.core.mul.Mul
```

**Comment by Arpan612** [comment]
Sir I mean KroneckerDelta(0, 1) is represented something likePiecewise((x, x < 1), (0, True), (1, True), correct? How will KroneckerDelta(n, 0) * cos(n) be represented?

File: sympy/functions/special/tests/test_tensor_functions.py [hunk 1]
```diff
@@ -57,6 +60,8 @@ def test_kronecker_delta():
     assert adjoint(KroneckerDelta(i, j)) == KroneckerDelta(i, j)
     assert conjugate(KroneckerDelta(i, j)) == KroneckerDelta(i, j)
     assert transpose(KroneckerDelta(i, j)) == KroneckerDelta(i, j)
+    assert KroneckerDelta(n, 0)*cos(n) == KroneckerDelta(n, 0)
```
Comment (sylee957) [inline sympy/functions/special/tests/test_tensor_functions.py]: The reason this is not failing is that you're using `n` from the above loop. 
So this test formulation would not be correct. You should define a symbol 'n' before

**Comment by Arpan612** [comment]
@mentors Could someone tell me why is the assert still failing?

**Comment by sylee957** [comment]
> @mentors Could someone tell me why is the assert still failing?

It's because you have not made any real patch to the issue yet.

> Sir I mean KroneckerDelta(0, 1) is represented something likePiecewise((x, x < 1), (0, True), (1, True), correct? How will KroneckerDelta(n, 0) * cos(n) be represented?

`KroneckerDelta(a, b)` is same as `Piecewise((0, Ne(a, b)), (1, True))`, so your formula is not correct.

And if you want to see how `KroneckerDelta(n, 0) * cos(n)` is represented, you should use 
`print_tree(KroneckerDelta(n, 0) * cos(n), evaluate=False)` or dotprint. You can see it's tree form
```
Mul: cos(n)*KroneckerDelta(0, n)
+-KroneckerDelta: KroneckerDelta(0, n)
| +-Zero: 0
| +-Symbol: n
+-cos: cos(n)
  +-Symbol: n
```
And I'd defer you to the tutorial https://docs.sympy.org/dev/tutorial/manipulation.html

**Comment by Arpan612** [comment]
@sylee957 Okay sir. On it.

**Comment by Arpan612** [comment]
@sylee957 Could you help me identify changes that need to be made in order to add the tests?
I was thinking of adding the following in test_simplify.py:
`assert simplify(K(0, j) * cos(n)) == K(0, j)`
 `assert simplify(K(0, j) * binomial(n, 2)) == K(0, j)  `

**Comment by Arpan612** [comment]
@sylee957 Could you please have a look?

**Comment by oscarbenjamin** [comment]
There isn't any code here to solve the issue so I'm going to close this now. If you want to open a new pull request you can but first see that you have some code that implements a change in sympy. If you want help with that then you can ask at the issue #18879.

**Pull request closed without merging**